packages/data-broker/src/DataBroker.js

```diff
@@ -63,20 +63,29 @@ export class DataBroker {
 
   async pull(dataSourceSpec, options) {
     const dataSources = await this.fetchDataSources(dataSourceSpec);
-    let results;
-    const pullForServiceAndType = async dataSourceGroup => {
-      const { type, service_type: serviceType } = dataSourceGroup[0];
-      const service = this.createService(serviceType);
-      const resultsForGroup = await service.pull(dataSourceGroup, type, options);
-      const mergeResults = this.resultMergers[type];
-      results = mergeResults(results, resultsForGroup);
-    };
+    if (countDistinct(dataSources, 'type') > 1) {
+      throw new Error('Cannot pull multiple types of data in one call');
+    }
 
-    const dataSourceGroups = Object.values(groupBy(dataSources, 'service_type'));
-    await Promise.all(dataSourceGroups.map(pullForServiceAndType));
-    return results;
+    const dataSourcesByService = groupBy(dataSources, 'service_type');
+    const nestedResults = await Promise.all(
+      Object.values(dataSourcesByService).map(dataSourcesForService =>
+        this.pullForServiceAndType(dataSourcesForService, options),
+      ),
+    );
+    const mergeResults = this.resultMergers[dataSources[0].type];
+
+    return nestedResults.reduce((results, resultsForService) =>
+      mergeResults(results, resultsForService),
+    );
   }
 
+  pullForServiceAndType = async (dataSources, options) => {
+    const { type, service_type: serviceType } = dataSources[0];
```

Comment:
Remind me: is validating that all are of the same `service_type` handled somewhere else?

Reply:
Yeah this "private" method is only called from `DataBroker.pull`, which creates groups of data sources per service type. In other words, this method's "contract" is that the input must be for a specific service and type.